### ISSUE TITLE
[Bug] A non-numeric value encountered

### ISSUE THREAD

**Issue by ajira86** [DirectoryTree/LdapRecord#457]
**Environment:**
 - LDAP Server Type: ActiveDirectory
 - PHP Version: 8.1

**Describe the bug:**

Getting LDAP user sometimes return this exception since I updated LDAP records from 2.11 to 2.14

```
{
    "message": "A non-numeric value encountered",
    "exception": "ErrorException",
    "file": "/var/www/vendor/directorytree/ldaprecord/src/Utilities.php",
    "line": 158,
    "trace": [
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php",
            "line": 255,
            "function": "handleError",
            "class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Utilities.php",
            "line": 158,
            "function": "Illuminate\\Foundation\\Bootstrap\\{closure}",
            "class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Attributes/Timestamp.php",
            "line": 229,
            "function": "convertWindowsTimeToUnixTime",
            "class": "LdapRecord\\Utilities",
            "type": "::"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Attributes/Timestamp.php",
            "line": 144,
            "function": "convertWindowsIntegerTimeToDateTime",
            "class": "LdapRecord\\Models\\Attributes\\Timestamp",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Concerns/HasAttributes.php",
            "line": 349,
            "function": "toDateTime",
            "class": "LdapRecord\\Models\\Attributes\\Timestamp",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Concerns/HasAttributes.php",
            "line": 140,
            "function": "asDateTime",
            "class": "LdapRecord\\Models\\Model",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Concerns/HasAttributes.php",
            "line": 100,
            "function": "addDateAttributesToArray",
            "class": "LdapRecord\\Models\\Model",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Concerns/SerializesAndRestoresPropertyValues.php",
            "line": 22,
            "function": "attributesToArray",
            "class": "LdapRecord\\Models\\Model",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/directorytree/ldaprecord/src/Models/Concerns/SerializesProperties.php",
            "line": 86,
            "function": "getSerializedPropertyValue",
            "class": "LdapRecord\\Models\\Model",
            "type": "->"
        },
        {
            "function": "__serialize",
            "class": "LdapRecord\\Models\\Model",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Session/Store.php",
            "line": 170,
            "function": "serialize"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php",
            "line": 236,
            "function": "save",
            "class": "Illuminate\\Session\\Store",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php",
            "line": 130,
            "function": "saveSession",
            "class": "Illuminate\\Session\\Middleware\\StartSession",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php",
            "line": 64,
            "function": "handleStatefulRequest",
            "class": "Illuminate\\Session\\Middleware\\StartSession",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Session\\Middleware\\StartSession",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php",
            "line": 37,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php",
            "line": 67,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Cookie\\Middleware\\EncryptCookies",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 116,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 726,
            "function": "then",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 703,
            "function": "runRouteWithinStack",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 667,
            "function": "runRoute",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 656,
            "function": "dispatchToRoute",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 167,
            "function": "dispatch",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 141,
            "function": "Illuminate\\Foundation\\Http\\{closure}",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php",
            "line": 60,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Barryvdh\\Debugbar\\Middleware\\InjectDebugbar",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
            "line": 21,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php",
            "line": 31,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
            "line": 21,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php",
            "line": 40,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TrimStrings",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php",
            "line": 27,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php",
            "line": 86,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php",
            "line": 49,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Http\\Middleware\\HandleCors",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php",
            "line": 39,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 180,
            "function": "handle",
            "class": "Illuminate\\Http\\Middleware\\TrustProxies",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 116,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 142,
            "function": "then",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 111,
            "function": "sendRequestThroughRouter",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        },
        {
            "file": "/var/www/public/index.php",
            "line": 52,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        }
    ]
}
```

**Comment by ajira86** [comment]
Seams to be an issue using `convertWindowsTimeToUnixTime` on `lastlogon` from active directory user attribute

**Comment by stevebauman** [comment]
Hi @ajira86, did you update to PHP 8.1 recently as well? This may be an 8.1 issue due to strict types.

**Comment by ajira86** [comment]
Nope, the project starts with PHP 8.1 about three month ago.

**Comment by stevebauman** [comment]
Ok, can you share your code and a dump of the model with any sensitive attributes omitted?

**Comment by ajira86** [comment]
```
    /**
     * Round a Windows timestamp down to seconds and remove
     * the seconds between 1601-01-01 and 1970-01-01.
     *
     * @param float $windowsTime
     *
     * @return float
     */
    public static function convertWindowsTimeToUnixTime($windowsTime)
    {
        dd($windowsTime);
        return round($windowsTime / 10000000) - 11644473600;
    }
```

I get sometimes `"132983861988617987"` or sometimes `"2022-05-30T14:09:59+0200"` in `$windowsTime` parameter

**Comment by stevebauman** [comment]
Can you share your code? Have you extended any models? etc.

**Comment by ajira86** [comment]
And here example of user on issue :

```
App\Ldap\User {#1623
  +exists: true
  +wasRecentlyCreated: false
  +wasRecentlyRenamed: false
  #dn: "CN=XXXXXXXXXX,OU=XXXXXXXXXX-XXXXXXXXXX,OU=XXXXXXXXXX,OU=Users,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
  #in: null
  #connection: null
  #guidKey: "objectguid"
  #modifications: []
  #original: array:41 [
    "objectclass" => array:4 [
      0 => "top"
      1 => "person"
      2 => "organizationalPerson"
      3 => "user"
    ]
    "cn" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "sn" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "telephonenumber" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "givenname" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "distinguishedname" => array:1 [
      0 => "CN=XXXXXXXXXX,OU=XXXXXXXXXX-XXXXXXXXXX,OU=XXXXXXXXXX,OU=Users,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
    ]
    "instancetype" => array:1 [
      0 => "4"
    ]
    "whencreated" => array:1 [
      0 => "20180410192706.0Z"
    ]
    "whenchanged" => array:1 [
      0 => "20220530135231.0Z"
    ]
    "usncreated" => array:1 [
      0 => "1539365"
    ]
    "memberof" => array:3 [
      0 => "CN=7Media XXXXXXXXXX XXXXXXXXXX,OU=Units,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
      1 => "CN=7Media XXXXXXXXXX XXXXXXXXXX Admin Group,OU=Units,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
      2 => "CN=Utilisateur Affilié,OU=Roles,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
    ]
    "usnchanged" => array:1 [
      0 => "4410957"
    ]
    "name" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "objectguid" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "useraccountcontrol" => array:1 [
      0 => "66080"
    ]
    "badpwdcount" => array:1 [
      0 => "0"
    ]
    "codepage" => array:1 [
      0 => "0"
    ]
    "countrycode" => array:1 [
      0 => "0"
    ]
    "badpasswordtime" => array:1 [
      0 => "132983922103667934"
    ]
    "lastlogoff" => array:1 [
      0 => "0"
    ]
    "lastlogon" => array:1 [
      0 => "132983940638135952"
    ]
    "pwdlastset" => array:1 [
      0 => "132948418776304854"
    ]
    "primarygroupid" => array:1 [
      0 => "513"
    ]
    "objectsid" => array:1 [
      0 => """
        \x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00\n
        ¡¡1KHÝ(\x12B(\x03;4\x00\x00
        """
    ]
    "accountexpires" => array:1 [
      0 => "9223372036854775807"
    ]
    "logoncount" => array:1 [
      0 => "342"
    ]
    "samaccountname" => array:1 [
      0 => "admin_XXXXXXXXXX"
    ]
    "samaccounttype" => array:1 [
      0 => "805306368"
    ]
    "userprincipalname" => array:1 [
      0 => "XXXXXXXXXX@XXXXXXXXXX.loc"
    ]
    "objectcategory" => array:1 [
      0 => "CN=Person,CN=Schema,CN=Configuration,DC=XXXXXXXXXX,DC=loc"
    ]
    "dscorepropagationdata" => array:4 [
      0 => "20180824080048.0Z"
      1 => "20180824075025.0Z"
      2 => "20180418141046.0Z"
      3 => "16010101000000.0Z"
    ]
    "lastlogontimestamp" => array:1 [
      0 => "132977632501342339"
    ]
    "mail" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "mobile" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "usersecuritytoken" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "sitefinityaccountauthmode" => array:1 [
      0 => "2"
    ]
    "sitefinityaccounttype" => array:1 [
      0 => "2"
    ]
    "sitefinityaccountauthprovider" => array:1 [
      0 => "2"
    ]
  ]
  #attributes: array:41 [
    "objectclass" => array:4 [
      0 => "top"
      1 => "person"
      2 => "organizationalPerson"
      3 => "user"
    ]
    "cn" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "sn" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "telephonenumber" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "givenname" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "distinguishedname" => array:1 [
      0 => "CN=XXXXXXXXXX,OU=XXXXXXXXXX-XXXXXXXXXX,OU=XXXXXXXXXX,OU=Users,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
    ]
    "instancetype" => array:1 [
      0 => "4"
    ]
    "whencreated" => array:1 [
      0 => "2018-04-10T19:27:06+0200"
    ]
    "whenchanged" => array:1 [
      0 => "2022-05-30T13:52:31+0200"
    ]
    "usncreated" => array:1 [
      0 => "1539365"
    ]
    "memberof" => array:3 [
      0 => "CN=7Media XXXXXXXXXX XXXXXXXXXX,OU=Units,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
      1 => "CN=7Media XXXXXXXXXX XXXXXXXXXX Admin Group,OU=Units,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
      2 => "CN=Utilisateur Affilié,OU=Roles,OU=XXXXXXXXXX,DC=XXXXXXXXXX,DC=loc"
    ]
    "usnchanged" => array:1 [
      0 => "4410957"
    ]
    "name" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "objectguid" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "useraccountcontrol" => array:1 [
      0 => "66080"
    ]
    "badpwdcount" => array:1 [
      0 => "0"
    ]
    "codepage" => array:1 [
      0 => "0"
    ]
    "countrycode" => array:1 [
      0 => "0"
    ]
    "badpasswordtime" => array:1 [
      0 => "2022-05-30T15:50:10+0200"
    ]
    "lastlogoff" => array:1 [
      0 => "0"
    ]
    "lastlogon" => array:1 [
      0 => "2022-05-30T16:21:04+0200"
    ]
    "pwdlastset" => array:1 [
      0 => "2022-04-19T13:37:58+0200"
    ]
    "primarygroupid" => array:1 [
      0 => "513"
    ]
    "objectsid" => array:1 [
      0 => """
        \x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00\n
        ¡¡1KHÝ(\x12B(\x03;4\x00\x00
        """
    ]
    "accountexpires" => array:1 [
      0 => "2008-09-14T04:48:05+0200"
    ]
    "logoncount" => array:1 [
      0 => "342"
    ]
    "samaccountname" => array:1 [
      0 => "admin_XXXXXXXXXX"
    ]
    "samaccounttype" => array:1 [
      0 => "805306368"
    ]
    "userprincipalname" => array:1 [
      0 => "XXXXXXXXXX@XXXXXXXXXX.loc"
    ]
    "objectcategory" => array:1 [
      0 => "CN=Person,CN=Schema,CN=Configuration,DC=XXXXXXXXXX,DC=loc"
    ]
    "dscorepropagationdata" => array:1 [
      0 => "2018-08-24T08:00:48+0200"
    ]
    "lastlogontimestamp" => array:1 [
      0 => "2022-05-23T09:07:30+0200"
    ]
    "mail" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "mobile" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "usersecuritytoken" => array:1 [
      0 => "XXXXXXXXXX"
    ]
    "sitefinityaccountauthmode" => array:1 [
      0 => "2"
    ]
    "sitefinityaccounttype" => array:1 [
      0 => "2"
    ]
    "sitefinityaccountauthprovider" => array:1 [
      0 => "2"
    ]
  ]
  #dates: array:7 [
    "lastlogon" => "windows-int"
    "lastlogoff" => "windows-int"
    "pwdlastset" => "windows-int"
    "lockouttime" => "windows-int"
    "accountexpires" => "windows-int"
    "badpasswordtime" => "windows-int"
    "lastlogontimestamp" => "windows-int"
  ]
  #casts: []
  #appends: []
  #dateFormat: null
  #defaultDates: array:3 [
    "whenchanged" => "windows"
    "whencreated" => "windows"
    "dscorepropagationdata" => "windows"
  ]
  #hidden: []
  #visible: []
  #sidKey: "objectsid"
  #passwordAttribute: "unicodepwd"
  #passwordHashMethod: "encode"
}
```

**Comment by stevebauman** [comment]
Thanks for posting all of that @ajira86, much appreciated.

Can you remove the trait named `AttributesForJson` on the `Ldap\User` model and try again?

This is likely due to the serialization bug fixes in [v2.12.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v2.12.0). I'm going to investigate further, but let me know if removing that trait resolves this serialization issue.

**Comment by stevebauman** [comment]
Patch is in the works -- will be released later today 👍

**Comment by ellisonpatterson** [comment]
> Patch is in the works -- will be released later today 👍

I'm experiencing this as well, I'm wondering if you had time to get that commit pushed.

Thank you so much!

**Comment by stevebauman** [comment]
Whoops! Thanks for the reminder @ellisonpatterson -- I forgot to push the PR 😅 

https://github.com/DirectoryTree/LdapRecord/pull/471

**Comment by stevebauman** [comment]
Ok I've just released [v2.15.0](https://github.com/DirectoryTree/LdapRecord/releaseshttps://github.com/DirectoryTree/LdapRecord/releases/tag/v2.15.0) with this patch. Let me know if you encounter any issues!

Apologies for the delay 😄

**Comment by ellisonpatterson** [comment]
No worries, thank you for the swift response and amazing work!
